src/model/misc.rs

```diff
@@ -74,6 +78,94 @@ impl Mentionable for User {
     }
 }
 
+#[cfg(feature = "cache")]
+impl FromStr for User {
+    type Err = ();
+    fn from_str(s: &str) -> StdResult<Self, ()> {
+        match utils::parse_username(s) {
+            Some(x) => {
+                match UserId(x as u64).find() {
+                    Some(user) => Ok(user),
+                    _ => Err(())
+                }
+            },
+            _ => Err(())
+        }
+    }
+}
+
+impl FromStr for UserId {
+    type Err = ();
+    fn from_str(s: &str) -> StdResult<Self, ()> {
+        match utils::parse_username(s) {
+            Some(x) => Ok(UserId(x)),
+            _ => Err(())
+        }
```

Comment:
Since this is only one level deep it can be chained into `utils::parse_username(s).ok_or_else(()).map(UserId)`. It gets messier to read when it's two levels deep like in the `FromStr` impl for `User`, so that doesn't need to be done above